code/site/controllers/http.php

```diff
@@ -64,15 +74,61 @@ public function display($cachable = false, $urlparams = array())
 	 *
 	 * @since 1.0
 	 */
-	private function sendError($exception)
+	private function respond($response)
 	{
-		JResponse::setHeader('status', $exception->getCode());
-		$error = new APIException($exception->getMessage(), $exception->getCode());
-		JFactory::getDocument()->setMimeEncoding('application/json');
+		$app = JFactory::getApplication();
+		$doc = JFactory::getDocument();
+		$accept = $app->input->server->get('HTTP_ACCEPT', 'application/json', 'STRING');
+		$compatibility = $app->input->server->get('HTTP_X_COMPATIBILITY_MODE', 0, 'INT');
+		
+		// Enforce JSON in compatibility mode
+		if ($compatibility)
+		{
+			$output = new \stdClass;
+			header("Content-type: application/json");
+			if ($response instanceof Exception) {
+				$output->message = $response->getMessage();
+				$output->code = $response->getCode();
+			}
+			else
+			{
+				$output = $response->get('response');
+			}
+
+			echo json_encode($output);
+
+			die();
+		}
+
+		switch($accept) {
+			case 'application/json':
+			default:
+				header("Content-type: application/json");
+				$format = 'json';
+			break;
+
+			case 'application/xml':
+				header("Content-type: application/xml");
+				$format = 'xml';
+			break;
+		}
+
+		$output_overrride = JPATH_ROOT . '/' . $app->getTemplate() . '/'.$format.'/api.php';
+
+		if (file_exists($output_overrride)) {
+			require_once $output_overrride;
+		}
+		else
+		{
+			require_once JPATH_COMPONENT . '/libraries/response/'.$format.'response.php';
+		}
+
+		$classname = 'API'.ucfirst($format).'Response';
+		$output = new $classname($response);
 
-		$error->newFormat = 1;
+		echo $output->__toString();
 
-		return json_encode($error->toArray());
+		die();
```

Comment:
Using jexit() was forcing the content-type header to be text/html so I've used die.